File: Algorithms/p187_Repeated_DNA_Sequences/p187_Repeated_DNA_Sequences_1.py
class Solution(object):
    def findRepeatedDnaSequences(self, s):
        """
        :type s: str
        :rtype: List[str]
        """
        res = set()
        seen = set()
        for i in range(10,len(s)+1):
            s1 = s[i-10:i]
            key = 0
            for c in s1:
                key <<= 2
                if c=='A':
                    key |= 0x0
                elif c=='C':
                    key |= 0x1
                elif c=='G':
                    key |= 0x2
                elif c=='T':
                    key |= 0x3
            if key in seen:
                res.add(s1)
            else:
                seen.add(key)
        return list(res)

File: Algorithms/p187_Repeated_DNA_Sequences/p187_Repeated_DNA_Sequences_1.cpp
class Solution {
public:
    vector<string> findRepeatedDnaSequences(string s) {
        vector<string> res;
        unordered_map<int,int> cnt;
        for(int i=10;i<=s.size();++i){
            int key = 0;
            for(int j=i-10;j<i;++j){
                key <<= 2;
                switch(s[j]){
                    case 'A': key |= 0x0; break;
                    case 'C': key |= 0x1; break;
                    case 'G': key |= 0x2; break;
                    case 'T': key |= 0x3; break;
                }
            }
            ++cnt[key];
            if(cnt[key]==2) res.push_back(s.substr(i-10,10));
        }
        return res;
    }
};
